### DISCUSSION TITLE
How to query for direct dependencies (components) for a project in the DB?

### DISCUSSION THREAD

**Discussion by walterdeboer** [DependencyTrack/dependency-track#2554] [Q&A] [answered]
I'm looking in the database to find direct dependencies for a project. I was expecting to find them by looking for components without parent_component_id, but I also find many transitive dependencies? How can I find only the direct dependencies for a project in the DB?

select * from component where parent_component_id is null and project_id = 1

**Comment by stevespringett** [comment, marked as answer]
Direct dependencies should be in component.directDependencies. It will be a JSON blob that you can parse through.

```sql
SELECT DIRECT_DEPENDENCIES FROM COMPONENT WHERE ID = 1;
```

**Reply by walterdeboer** [reply]
Ah, I see, and also in PROJECT: 

SELECT DIRECT_DEPENDENCIES FROM PROJECT TWHERE ID = 1;

No easy way to query the components from the JSON blob then, I guess? I'll see how to work my way around that...

Thanks!

**Reply by walterdeboer** [reply]
Would it be an idea to fix the DB so parent_component_id could be used? It looks like it serves this pupose allready from some components? Why not for all?

It would certainly make life easier for my outdated components query:

https://github.com/walterdeboer/dependency-track/blob/feature/2514/src/main/java/org/dependencytrack/model/Finding.java#L134

I'd rather not include indirect dependencies there...

**Reply by stevespringett** [reply]
No, parent/child components are for describing a component being included inside another component, which is very different than a dependency.

In CycloneDX, we call this nested relationship an assembly.
https://cyclonedx.org/use-cases/#assembly

**Reply by walterdeboer** [reply]
Oh, I see. This should do the trick then, thanks!

```
select * 
from component, project
where component.project_id = project.id
and project.id = 1
and project.DIRECT_DEPENDENCIES like CONCAT('%"uuid":"', component.uuid, '"%')
order by component.purl 
```
